File: leetcode/0199-binary-tree-right-side-view/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        res = []
        q = []

        q.append(root)

        while q:
            level_size = len(q)
            level = []
            while level_size > 0:
                el = q.pop(0)
                level.append(el.val)
                if el.left:
                    q.append(el.left)
                if el.right:
                    q.append(el.right)
                level_size -= 1
            res.append(level[-1])
        
        return res

File: leetcode/0199-binary-tree-right-side-view/solution.java
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if (root == null)
            return new ArrayList();
        Queue<TreeNode> queue = new LinkedList();
        queue.offer(root);
        List<Integer> res = new ArrayList();
        
        while(!queue.isEmpty()){
            int size = queue.size();
            
            while (size -- > 0){
                TreeNode cur = queue.poll();
                if (size == 0)
                    res.add(cur.val);
                
                if (cur.left != null)
                    queue.offer(cur.left);
                if (cur.right != null)
                    queue.offer(cur.right);
            }
        }
        
        return res;
    }
}
